replica/database.cc

```diff
@@ -2008,6 +2010,7 @@ future<> database::do_apply(schema_ptr s, const frozen_mutation& m, tracing::tra
             co_await coroutine::return_exception_ptr(std::move(ex));
         }
         lock = lock_f.get();
+        backlog = _view_update_generator->get_proxy().get_view_update_backlog();
```

Comment:
> can't we have have the code that needs it call _view_update_generator->get_proxy().get_view_update_backlog() to get it?

This is what the code does today and this patch fixes it like this

```patch
@@ -4070,10 +4075,10 @@ void storage_proxy::send_to_live_endpoints(storage_proxy::response_id_type respo
     // lambda for applying mutation locally
     auto lmutate = [handler_ptr, response_id, this, my_address, timeout] () mutable {
         return handler_ptr->apply_locally(timeout, handler_ptr->get_trace_state())
-                .then([response_id, this, my_address, h = std::move(handler_ptr), p = shared_from_this()] {
+                .then([response_id, this, my_address, h = std::move(handler_ptr), p = shared_from_this()] (std::optional<db::view::update_backlog> b) {
             // make mutation alive until it is processed locally, otherwise it
             // may disappear if write timeouts before this future is ready
-            got_response(response_id, my_address, get_view_update_backlog());
+            got_response(response_id, my_address, b);
         });
     };
```

so your proposal is to leave things as is :)

Reply:
> ... so your proposal is to leave things as is :)

And I second that suggestion, by the way

Reply:
> That's horrible layering violation :( Can view_update_generator::push_view_replica_updates() call _proxy.get_view_update_backlog() on its own and return it back to database (and then up the chain) at least?

`push_view_replica_updates()` is actually a method of `table` so I suppose we don't want to refer to the storage_proxy there either, so I moved it to `view_update_generator::generate_and_propagate_view_updates`



> I would ask for the almost opposite thing, which I think will remove 99% of this patch:
> 
> Instead of having all this code return the "view update backlog" and changing many functions to pass this backlog up the chain, can't we have have the code that needs it call _view_update_generator->get_proxy().get_view_update_backlog() to get it?

Instead of this entire chain we could have the code that needs it call `_db.invoke_on(..., get_view_update_backlog())` but I expect we don't want to add it to the common write path due to performance considerations

Reply:
Since storage_proxy::get_view_update_backlog() is using atomics to fetch-add its numbers, it can be safely used from _any_ shard. So instead of propagating the result of this function all over the code, we (probably) can teach proxy to call its get_view_update_backlog() with int shard argument, this won't cost us extra submit-to-s

Reply:
`storage_proxy::get_view_update_backlog()` is using atomics to store and read the cached backlogs, but the value to be stored in the atomics comes from `database::get_view_update_backlog` which reads the semaphore of the shard it's called on, so to use that we would need the `invoke_on`

Reply:
> Since storage_proxy::get_view_update_backlog() is using atomics to fetch-add its numbers, it can be safely used from any shard. So instead of propagating the result of this function all over the code, we (probably) can teach proxy to call its get_view_update_backlog() with int shard argument, this won't cost us extra submit-to-s

@xemul Each shard's atomic is currently written only by its shard and occasionally (every 10ms) read by other shards. Consider a non-shard-aware workload - won't this have a significant performance impact if multiple shards will be allowed to write to the same atomic on each write operation?